File: apps/iosApp/iosApp/ar/ARBridge.swift
import Foundation
import Combine

/// Bridge between KMP shared navigation state and native iOS AR layer.
///
/// Connects the shared ArNavigationCoordinator to the native iOS AR
/// components (session manager, marker detector, route renderer).
///
/// TODO: Import shared KMP framework once iOS build is configured.
/// For now, this follows the ArNavigationBridge interface contract.
class ARBridge: ObservableObject {
    
    /// Human-readable label for the current session state.
    @Published var currentStateLabel: String = "Idle"
    
    /// Whether the entrance marker has been detected.
    @Published var isWorldAligned: Bool = false
    
    /// Current navigation instruction text.
    @Published var currentInstruction: String = ""
    
    /// Number of arrows currently renderable.
    @Published var arrowCount: Int = 0
    
    /// Current destination name.
    @Published var destinationName: String = ""
    
    // MARK: - ArNavigationBridge Implementation
    
    /// Start an AR session. Called by shared coordinator.
    /// TODO: Accept BuildingPackage and Room from KMP shared code.
    func startSession(buildingId: String, destinationName: String) {
        currentStateLabel = "Waiting for Marker"
        self.destinationName = destinationName
        currentInstruction = "Point camera at the entrance marker"
    }
    
    /// Stop the AR session.
    func stopSession() {
        currentStateLabel = "Idle"
        isWorldAligned = false
        currentInstruction = ""
        arrowCount = 0
    }
    
    /// Called when alignment is established.
    func onAlignmentEstablished(
        offsetX: Double, offsetY: Double, offsetZ: Double,
        rotationYDeg: Double
    ) {
        isWorldAligned = true
        currentStateLabel = "Aligned"
    }
    
    /// Called when renderable route is updated.
    func onRenderableRouteUpdated(arrowCount: Int, instruction: String) {
        self.arrowCount = arrowCount
        currentInstruction = instruction
        currentStateLabel = "Rendering Route"
    }
    
    /// Called when session state changes.
    func onSessionStateChanged(stateLabel: String) {
        currentStateLabel = stateLabel
    }
}

File: apps/androidApp/src/main/kotlin/com/vecturai/android/ar/ArBridge.kt
package com.vecturai.android.ar

import com.vecturai.core.domain.NavigationState
import com.vecturai.core.store.AppStore
import kotlinx.coroutines.flow.StateFlow

/**
 * Bridge between shared KMP AR coordinator and native Android AR layer.
 *
 * Follows the shared ArNavigationBridge interface contract.
 * Connects ArNavigationCoordinator events to the native AR components.
 *
 * TODO: Wire to ArNavigationCoordinator via Koin DI once shared framework builds.
 */
class ArBridge {

    /** Current session state label (for debug display). */
    var currentStateLabel: String = "Idle"
        private set

    /** Whether marker alignment is established. */
    var isWorldAligned: Boolean = false
        private set

    /** Current instruction text. */
    var currentInstruction: String = ""
        private set

    /** Arrow count currently renderable. */
    var arrowCount: Int = 0
        private set

    // ── ArNavigationBridge contract ────────────────

    fun startSession(buildingId: String, destinationName: String) {
        currentStateLabel = "Waiting for Marker"
        currentInstruction = "Point camera at the entrance marker"
    }

    fun stopSession() {
        currentStateLabel = "Idle"
        isWorldAligned = false
        currentInstruction = ""
        arrowCount = 0
    }

    fun onAlignmentEstablished(
        offsetX: Double, offsetY: Double, offsetZ: Double,
        rotationYDeg: Double,
    ) {
        isWorldAligned = true
        currentStateLabel = "Aligned"
    }

    fun onRenderableRouteUpdated(arrowCount: Int, instruction: String) {
        this.arrowCount = arrowCount
        currentInstruction = instruction
        currentStateLabel = "Rendering Route"
    }

    fun onSessionStateChanged(stateLabel: String) {
        currentStateLabel = stateLabel
    }

    /**
     * Observe shared navigation state.
     */
    fun observeNavigationState(appStore: AppStore): StateFlow<NavigationState> {
        return appStore.navigationState
    }

    /**
     * Called when entrance marker is detected by ARCore.
     */
    fun onEntranceMarkerDetected(markerPayload: String) {
        isWorldAligned = true
        currentStateLabel = "Navigating"
    }

    fun onDestinationReached() {
        currentStateLabel = "Arrived"
    }
}
